### ISSUE TITLE
RMSprop_Graves implementation issue

### ISSUE THREAD

**Issue by jamesthomasgriffin** [chainer/chainer#3451]
There is an issue regarding the implementation of the RMSpropGraves optimiser, I've included the relevant code from `rmsprop_graves.py` below.  In this optimiser the gradient of each parameter is rescaled by the standard deviation, calculated using a running average of the square of the gradient.  The issue is that the running averages are being computed incorrectly because the optimiser is only given the average of each batch of gradients, rather than the individual gradients. 

The consequence is that the estimator of the standard deviation is a factor of sqrt(N) smaller than it should be (where N is the batch size), and the error associated to this estimate will be somewhat larger.  The factor of sqrt(N) is not a major issue, it has the effect of rescaling the learning rate.  The increased error is an issue and will lead to increased volatility of the gradient.

Looking back to the original paper [https://arxiv.org/pdf/1308.0850.pdf](url) it doesn't mention using batches at all.  A quick Google didn't turn up implementations in other frameworks.

I would be happy to contribute code if someone could tell me how to:

a) make each individual gradient of the batch available to the optimiser, or
b) compute the average squares per batch elsewhere and then pass this to the optimiser.

```
def update_core_cpu(self, param):
        grad = param.grad
        if grad is None:
            return
        n, g, delta = self.state['n'], self.state['g'], self.state['delta']
        hp = self.hyperparam
        n *= hp.alpha
        n += (1 - hp.alpha) * grad * grad
        g *= hp.alpha
        g += (1 - hp.alpha) * grad
        delta *= hp.momentum
        delta -= hp.lr * grad / numpy.sqrt(n - g * g + hp.eps)
        param.data += delta
```
should in principal be
```
def update_core_cpu(self, param):
        grads = param.batch_of_grads   # Obviously doesn't work, but suppose it did
        if grads is None:
            return
        grad = np.mean(grads, axis=0)
        n, g, delta = self.state['n'], self.state['g'], self.state['delta']
        hp = self.hyperparam
        n *= hp.alpha
        n += (1 - hp.alpha) * np.mean(grads * grads, axis=0)
        g *= hp.alpha
        g += (1 - hp.alpha) * grad
        delta *= hp.momentum
        delta -= hp.lr * grad / numpy.sqrt(n - g * g + hp.eps)
        param.data += delta
```

NB  For standard RMSprop, there is a similar scaling by a running average of squares, however this is as a proxy for the magnitude of the gradient, so the current implementation is probably fine.

**Comment by muupan** [comment]
@jamesthomasgriffin 

> A quick Google didn't turn up implementations in other frameworks.

Chainer's `RMSpropGraves` is implemented  in tensorflow  and pytorch as RMSprop with the "centered" option.

https://www.tensorflow.org/api_docs/python/tf/train/RMSPropOptimizer
http://pytorch.org/docs/master/optim.html#torch.optim.RMSprop

Computing individual gradients would be computationally much more expensive, so I guess their centered RMSprop don't use individual gradients either. Correct me if I'm wrong.
